modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+    {
+      node_id[i] = cut_elem->get_node(i)->id();
+      is_node_on_boundary[i] = (_boundary_map.find(node_id[i]) != _boundary_map.end());
+    }
+
+    if (is_node_on_boundary[0] && is_node_on_boundary[1] && is_node_on_boundary[2])
+    {
+      // store corner elements
+      corner_elem_id.push_back(counter);
+    }
+    else
+    {
+      // for other elements, store boundary edges
+      for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+      {
+        if (is_node_on_boundary[i] && is_node_on_boundary[(i + 1 <= 2) ? i + 1 : 0])
+        {
+          unsigned int node1 = node_id[i];
+          unsigned int node2 = node_id[(i + 1 <= 2) ? i + 1 : 0];
+          if (node1 > node2)
+            std::swap(node1, node2);
+          std::shared_ptr<CutEdge> ptr(new CutEdge);
+          _boundary_edges.push_back(ptr);
+          _boundary_edges[_boundary_edges.size() - 1]->id1 = node1;
+          _boundary_edges[_boundary_edges.size() - 1]->id2 = node2;
+        }
+      }
+    }
+  }
+
+  // loop over edges in corner elements
+  for (unsigned int i = 0; i < corner_elem_id.size(); ++i)
+  {
+    elem_it = elem_begin;
+    for (unsigned int j = 0; j < corner_elem_id[i]; ++j)
+      ++elem_it;
+    Elem * cut_elem = *elem_it;
+
+    for (unsigned int j = 0; j < _cut_elem_nnode; ++j)
+    {
+      bool is_edge_inside = 0;
+
+      unsigned int node1 = cut_elem->get_node(j)->id();
+      unsigned int node2 = cut_elem->get_node((j + 1 <= 2) ? j + 1 : 0)->id();
+      if (node1 > node2)
+        std::swap(node1, node2);
+
+      MeshBase::element_iterator elem_it2 = elem_begin;
+      int counter = -1;
+      for (; elem_it2 != elem_end; ++elem_it2)
+      {
+        ++counter;
+
+        if (counter != corner_elem_id[i])
+        {
+          Elem * cut_elem2 = *elem_it2;
+          for (unsigned int k = 0; k < _cut_elem_nnode; ++k)
+          {
+            unsigned int node3 = cut_elem2->get_node(k)->id();
+            unsigned int node4 = cut_elem2->get_node((k + 1 <= 2) ? k + 1 : 0)->id();
+            if (node3 > node4)
+              std::swap(node3, node4);
+
+            if (node1 == node3 && node2 == node4)
+            {
+              is_edge_inside = 1;
+              goto endloop;
+            }
+          }
+        }
+      }
+    endloop:
+      if (is_edge_inside == 0)
+      {
+        // store good edges
+        std::shared_ptr<CutEdge> ptr(new CutEdge);
+        _boundary_edges.push_back(ptr);
+        _boundary_edges[_boundary_edges.size() - 1]->id1 = node1;
+        _boundary_edges[_boundary_edges.size() - 1]->id2 = node2;
+      }
+      else
+      {
+        // this is a bad edge; remove it from existing edge list
+        for (unsigned int k = 0; k < _boundary_edges.size(); ++k)
+        {
+          if (_boundary_edges[k]->id1 == node1 && _boundary_edges[k]->id2 == node2)
+          {
+            _boundary_edges.erase(_boundary_edges.begin() + k);
```

Comment:
Using operator+() on iterators isn't good if we switch containers. You should just use iterators for this loop.